improver/regrid/idw.py

```diff
@@ -0,0 +1,218 @@
+    )
+
+    # Add a small amount to all distances to avoid division by zero when taking the inverse
+    masked_distances += np.finfo(np.float32).eps
+    # Invert the distances, sum the k surrounding points, scale to produce weights
+    inv_distances = 1.0 / masked_distances
+    # add power 1.80 for inverse diatance weight
+    optimum_power = 1.80
+    inv_distances_power = np.power(inv_distances, optimum_power)
+    inv_distances_sum = np.sum(inv_distances_power, axis=1)
+    inv_distances_sum = 1.0 / inv_distances_sum
+    weights_idw = inv_distances_power * inv_distances_sum.reshape(-1, 1)
+
+    # Update indexes and weights with new values
+    indexes[output_points_match] = indexes_updates[points_with_match]
+    weights[output_points_match] = weights_idw
+    return indexes, weights, output_points_no_match
+
+
+def nearest_input_pts(in_latlons, out_latlons, k):
+    """
+    Find k nearest source (input) points to each target (output)
+    point, using a KDtree
+
+    Args:
+        in_latlons (numpy.ndarray):
+            Source grid points' latitude-longitudes (N x 2)
+        out_latlons (numpy.ndarray):
+            Target grid points' latitude-longitudes (M x 2)
+        k (int):
+            Number of points surrounding each output point
+
+    Return:
+        Tuple[numpy.ndarray, numpy.ndarray]:
+            Distances from target grid point to source grid points and indexes
+            of those points (M x K)
+    """
+    # Convert input latitude and longitude to XYZ coordinates, then create KDtree
+    in_x, in_y, in_z = ecef_coords(in_latlons[:, 0].flat, in_latlons[:, 1].flat)
+    in_coords = np.c_[in_x, in_y, in_z]
+    in_kdtree = KDTree(in_coords)
+    # Convert output to XYZ and query the KDtree for nearby input points
+    out_x, out_y, out_z = ecef_coords(out_latlons[:, 0].flat, out_latlons[:, 1].flat)
+    out_coords = np.c_[out_x, out_y, out_z]
+    distances, indexes = in_kdtree.query(out_coords, k)
+    # Avoid single dimension output for k=1 case
+    if distances.ndim == 1:
+        distances = np.expand_dims(distances, axis=1)
+    if indexes.ndim == 1:
+        indexes = np.expand_dims(indexes, axis=1)
+    return distances, indexes
+
+
+def ecef_coords(lats, lons, alts=np.array(0.0)):
+    """
+    Transforms the coordinates to Earth Centred Earth Fixed coordinates
+    with WGS84 parameters. used in function _nearest_input_pts
+    Args:
+        lats(numpy.ndarray):
+            latitude coordinates
+        lons(numpy.ndarray):
+            longitude coordinates
+        alts(numpy.ndarray):
+            altitudes coordinates
+    Return:
+        Tuple[numpy.ndarray, numpy.ndarray, numpy.ndarray]:
+            Transformed coordinates
+    """
+    rlats = np.deg2rad(lats)
+    rlons = np.deg2rad(lons)
+    clats = np.cos(rlats)
+    clons = np.cos(rlons)
+    slats = np.sin(rlats)
+    slons = np.sin(rlons)
+    n = WGS84_A / np.sqrt(1.0 - (WGS84_E * WGS84_E * slats * slats))
```

Comment:
Why is the geoid hard coded here rather than taken from the input cube?

Reply:
It is actually copied from an existing code: preprocessing for calling KDtree.

Reply:
If you pass in the input cube you can access the coordinate system by doing
coord_system = cube.coord("latitude").coord_system
If it has the correct metadata I think it should return an iris,coord_systems.GeogCS instance  ( https://scitools-iris.readthedocs.io/en/stable/generated/api/iris/coord_systems.html#iris.coord_systems.GeogCS) that should have attributes you can interrogate to find all the geiod information. You can always default to WGS84 parameters if you don't find any on the cube. I'd be happy if this was done in a follow up PR based on where we are now if you want to make this change

Reply:
This will be dealt with in a follow on PR. Recorded on #1464